crates/engine/tree/src/download.rs

```diff
@@ -29,6 +46,241 @@ pub enum DownloadOutcome {
+{
+    /// Handles incoming download actions.
+    fn on_action(&mut self, event: DownloadAction) {
+        match event {
+            DownloadAction::Clear => self.clear(),
+            DownloadAction::Download(request) => self.download(request),
+        }
+    }
+
+    /// Advances the download process.
+    fn poll(&mut self, cx: &mut Context<'_>) -> Poll<DownloadOutcome> {
+        // advance all full block requests
+        for idx in (0..self.inflight_full_block_requests.len()).rev() {
+            let mut request = self.inflight_full_block_requests.swap_remove(idx);
+            if let Poll::Ready(block) = request.poll_unpin(cx) {
+                trace!(target: "consensus::engine", block=?block.num_hash(), "Received single full block, buffering");
+                self.set_buffered_blocks.push(Reverse(block.into()));
+            } else {
+                // still pending
+                self.inflight_full_block_requests.push(request);
+            }
+        }
+
+        // advance all full block range requests
+        for idx in (0..self.inflight_block_range_requests.len()).rev() {
+            let mut request = self.inflight_block_range_requests.swap_remove(idx);
+            if let Poll::Ready(blocks) = request.poll_unpin(cx) {
+                trace!(target: "consensus::engine", len=?blocks.len(), first=?blocks.first().map(|b| b.num_hash()), last=?blocks.last().map(|b| b.num_hash()), "Received full block range, buffering");
+                self.set_buffered_blocks.extend(
+                    blocks
+                        .into_iter()
+                        .map(|b| {
+                            let senders = b.senders().unwrap_or_default();
+                            OrderedSealedBlockWithSenders(SealedBlockWithSenders {
+                                block: b,
+                                senders,
+                            })
+                        })
+                        .map(Reverse),
+                );
+            } else {
+                // still pending
+                self.inflight_block_range_requests.push(request);
+            }
+        }
+
+        self.update_block_download_metrics();
+
+        // drain all unique element of the block buffer if there are any
+        let mut downloaded_blocks: Vec<SealedBlockWithSenders> =
+            Vec::with_capacity(self.set_buffered_blocks.len());
+        while let Some(block) = self.set_buffered_blocks.pop() {
+            // peek ahead and pop duplicates
+            while let Some(peek) = self.set_buffered_blocks.peek_mut() {
+                if peek.0 .0.hash() == block.0 .0.hash() {
+                    PeekMut::pop(peek);
+                } else {
+                    break
+                }
+            }
+            downloaded_blocks.push(block.0.into());
+        }
+        if downloaded_blocks.len() > 0 {
+            return Poll::Ready(DownloadOutcome::Blocks(downloaded_blocks))
+        }
```

Comment:
I believe we can flip this and move to L262

```rust
if self.set_buffered_blocks.is_empty() {
  return Poll::Pending
}
```

because if this is non empty we return at least 1 block, right?

Reply:
good point, done